File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.Stack;
class Solution {
    public int[] solution(int[] numbers) {
        int N = numbers.length;
        int[] answer = new int[N];
        Stack<Integer> stack = new Stack<>();
        stack.push(0);
        
        for (int i = 0; i < N; i++) {
            while (!stack.isEmpty() && numbers[stack.peek()] < numbers[i]) {
                answer[stack.pop()] = numbers[i];
            }
            stack.push(i);
        } 
        while (!stack.isEmpty()) {
            answer[stack.pop()] = -1;
        }
        return answer;
    }
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
def solution(numbers):
    answer = [0] * len(numbers)
    stack = []
    stack.append(0)
    for i in range(1, len(numbers)):
        while (stack and numbers[stack[-1]] < numbers[i]):
            answer[stack[-1]] = numbers[i]
            stack.pop()
        stack.append(i)
    while stack:
        answer[stack[-1]] = -1
        stack.pop()
    return answer